### PR TITLE
Performance > Duplicate queries for @mentions when rendering activity feed - Fixed

### LINKED ISSUE #3070: Performance issue - Lots of duplicate queries for @mention when activity feed is rendering
**Describe the bug**
In activity feed rendering there are 8-12 duplicate queries (depending on settings) for a single @mention which causing a huge performance drop when there are multiple @mention and make the activity feed page unuseable for large data site.

Please see this video for details and replication: https://www.loom.com/share/3d886ff7442e4bfa98c9da2d634e700c

**To Reproduce**
See the above video.
1. Create an activity post with multiple @mentions
2. Capture the queries for the activity item where has multiple @mentions when rendering.
3. See lots of duplicate queries.

**Expected behavior**
Duplicate queries should not run.

**Environment**
Client site, Dev, Local

**Support ticket links**
https://secure.helpscout.net/conversation/1645082023/166301/

Related Theme Issue: https://github.com/buddyboss/buddyboss-theme/issues/1690

### PR DESCRIPTION
**Pull request by mahedihasannoman**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Prevented initiating new activity object(BP_Activity_Activity) if the activity object is already available in the global activity template.

Fixes #3070 .

### How to test the changes in this Pull Request:

1. Create an activity post with multiple @mentions
2. Capture the queries for the activity item where has multiple @mentions when rendering.
3. See no duplicate queries

### Proof Screenshots or Video

https://www.loom.com/share/0a90b9669792476c8a0fa3245177a30b

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

**Review by dipskakadiya — DISMISSED** [review]

File: src/bp-activity/bp-activity-functions.php [hunk 1]
```diff
@@ -3541,10 +3541,22 @@ function bp_activity_delete_children( $activity_id, $comment_id ) {
...
 
 	if ( empty( $activity_obj ) ) {
-		$activity_obj = new BP_Activity_Activity( $activity_id );
+
+		/**
+		 * Check if activity object is available in global activity template.
+		 * If available then take the activity object from global template.
+		 * Else create a new instance.
+		 */
+		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $activity_id ) {
```
Comment (chetansatasiya) [inline src/bp-activity/bp-activity-functions.php:3554]:
```suggestion
		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $activity_id ) {
```

File: src/bp-forums/activity.php [hunk 2]
```diff
@@ -416,14 +416,30 @@ public function before_activity_content( $content, $activity ) {
...
-				return $buttons;
-			}
+			global $activities_template;
+
+			/**
+			 * Check if activity object is available in global activity template.
+			 * If available then take the activity object from global template.
+			 * Else create a new instance.
+			 */
+			if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $activity_id ) {
```
Comment (chetansatasiya) [inline src/bp-forums/activity.php:427]:
```suggestion
			if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $activity_id ) {
```

File: src/bp-moderation/classes/class-bp-moderation-activity.php [hunk 3]
```diff
@@ -97,7 +97,19 @@ public static function get_permalink( $activity_id ) {
...
-		$activity = new BP_Activity_Activity( $activity_id );
+
+		global $activities_template;
+
+		/**
+		 * Check if activity object is available in global activity template.
+		 * If available then take the activity object from global template.
+		 * Else create a new instance.
+		 */
+		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $activity_id ) {
```
Comment (chetansatasiya) [inline src/bp-moderation/classes/class-bp-moderation-activity.php:108]:
```suggestion
		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $activity_id ) {
```

File: src/bp-moderation/classes/class-bp-moderation-activity.php [hunk 4]
```diff
@@ -178,7 +190,18 @@ public function restrict_single_item( $restrict, $activity ) {
...
 
-		$activity = new BP_Activity_Activity( $item_id );
+		global $activities_template;
+
+		/**
+		 * Check if activity object is available in global activity template.
+		 * If available then take the activity object from global template.
+		 * Else create a new instance.
+		 */
+		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $item_id ) {
```
Comment (chetansatasiya) [inline src/bp-moderation/classes/class-bp-moderation-activity.php:200]:
```suggestion
		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $item_id ) {
```

File: src/bp-moderation/classes/class-bp-moderation-activity.php [hunk 5]
```diff
@@ -328,11 +351,23 @@ public function update_button_sub_items( $item_id ) {
...
 			return $retval;
 		}
 
-		$activity = new BP_Activity_Activity( (int) $item_id );
+		/**
+		 * Check if activity object is available in global activity template.
+		 * If available then take the activity object from global template.
+		 * Else create a new instance.
+		 */
+		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id == $item_id ) {
```
Comment (chetansatasiya) [inline src/bp-moderation/classes/class-bp-moderation-activity.php:366]:
```suggestion
		if ( isset( $activities_template ) && isset( $activities_template->activity ) && isset( $activities_template->activity->id ) && $activities_template->activity->id === $item_id ) {
```

**Review by chetansatasiya — CHANGES_REQUESTED** [review]
@mahedihasannoman Use the strict comparison in all the places in condition.

**Review by chetansatasiya — CHANGES_REQUESTED** [review]
@mahedihasannoman I debugged the issue and there is nothing wrong with these functions you made a change on this PR.

The issue is here https://github.com/buddyboss/buddyboss-platform/blob/release/src/bp-members/bp-members-functions.php#L307 as it's making a lot of queries to DB.

We can add the cache on this function like this.

```

function bp_core_get_userid_from_nickname( $nickname ) {
	if ( empty( $nickname ) ) {
		return false;
	}

	static $cache = array();

	$cache_key = 'bp_core_get_userid_from_nickname' . $nickname;
	if ( ! isset( $cache[ $cache_key ] ) ) {
		$cache[ $cache_key ] = get_users(
			array(
				'meta_key'    => 'nickname',
				'meta_value'  => $nickname,
				'number'      => 1,
				'count_total' => false,
			)
		);

		$user = $cache[ $cache_key ] ? $cache[ $cache_key ][0] : null;
	} else {
		$user = $cache[ $cache_key ] ? $cache[ $cache_key ][0] : null;
	}

	return apply_filters( 'bp_core_get_userid_from_nickname', ! empty( $user->ID ) ? $user->ID : null, $nickname );
}

```

**Comment by mahedihasannoman** [comment]
Hi @chetansatasiya , I have added cache on that function as you suggested and it is working fine. Also I have removed all of my previous code and conditions. Thanks.

**Review by chetansatasiya — APPROVED** [review]

**Review by KartikSuthar — APPROVED** [review]

**Comment by sumitbuddyboss** [comment]
Hello @mahedihasannoman 

The related issue is resolved, Now no more duplicate queries for @mention when the activity feed is rendering.

**Video proof:**
https://www.loom.com/share/9c8d80561a6b4892b5239cd41a163e6d

**Testcase:**
https://docs.google.com/spreadsheets/d/1v7bvYc-n-VZYQu8CcAcc9jLmHFduAQmxz_0n8lhjw7o/edit?pli=1#gid=1451538933

Ready for release 🚀 

Thank you

**Comment by sumitbuddyboss** [comment]
Hello @mahedihasannoman 

I have tested in a qa-approved branch and ready for release.

Test staus: PASS

Ready for release 🚀 

Thank you

**Pull request merged**